### ISSUE TITLE
xtext only detects comments immediately prior to the expression, not on the line above

### ISSUE THREAD

**Issue by cpeel** [vertilia/text#3]
`xgettext` will detect comments on the line before the translation string. Because `xtext` uses an actual PHP parser, it (appears to) only detects commands of the expression immediately before the translation string.

For example:
```php
// TRANSLATORS: say hello
$string = sprintf(_("Hello, world"));

$string = sprintf(
    // TRANSLATORS: bye bye bye
    _("Bye bye birdie")
);
```

`xgettext` will output (ignore the `php-format` -- it's technically accurate but not useful for this example):
```
#. TRANSLATORS: say hello
#: test.php:4
#, php-format
msgid "Hello, world"
msgstr ""

#. TRANSLATORS: bye bye bye
#: test.php:8
#, php-format
msgid "Bye bye birdie"
msgstr ""
```

`xtext` will output:
```
#: test.php:4
msgid "Hello, world"
msgstr ""

#. TRANSLATORS: bye bye bye
#: test.php:8
msgid "Bye bye birdie"
msgstr ""
```

I think this could be solved by keeping track of comments as they are encountered. When a string translation expression is encountered check if the line of the last comment (well, the last line of the prior comment) was on the line immediately above the string translation expression.

---

FYI: [Distributed Proofreaders](https://github.com/DistributedProofreaders/dproofreaders) is trying to use `xtext` to replace `xgettext` as our translation-prep tool primarily because `xgettext` is completely ignorant of PHP 7.3's indented heredoc. I realize that it might not be an objective for `xtext` to be a drop-in replacement for `xgettext` so please let me know if these issues and PRs are not wanted / welcome.

**Comment by dotgit** [comment]
I don't think it's realistic to completely replace the `xgettext` with a tool like `xtext`. Look at the LOC of each project.

The intended purpose was to ease the pain for the most useful functionality, like text extraction from heredoc strings, addition of context-related translation, decoupling from system locale.

Could you please provide additional use cases where the tool could be useful for your project? 

Meantime I will try to figure out the comments issue.

**Comment by cpeel** [comment]
> I don't think it's realistic to completely replace the `xgettext` with a tool like `xtext`. Look at the LOC of each project.

Agreed, and I'm not interested in supporting all of them. But I'd love for it to support just enough for us at DProofreaders 😁 

> Could you please provide additional use cases where the tool could be useful for your project?

Our initial testing says that `xtext` was just missing 3 things that we needed:
1. Better `%` detection / support for `// xgettext:no-php-format` string. The PR you merged seems to have solved our current `%` strings without the need for the comment-detection. 👍 
2. Support for translators notes on the line before, not expression before (this issue). And frankly this is something we could go through our codebase and fix if necessary.
3. Support for extracting strings from user-defined `pgettext()` functions (https://github.com/vertilia/text/pull/4)

> Meantime I will try to figure out the comments issue.

I'm also happy to try and figure this out myself and submit PRs too, just let me know!


I love that `text` exists and it sure would have saved us a lot of hassle on getting gettext working many years ago -- dealing with local locales is a huge pain in the ass, as you know. In transparency I'm not certain we'll transition our codebase to using the runtime component of `text` due to the lift of changing all of the calls, but I am hoping we can move to using `xtext` for string extraction as `xgettext` is simply not usable as-is.

Thank you for your time and energy!

**Comment by dotgit** [comment]
2. and 3. added in 1.5.0

**Comment by cpeel** [comment]
Thanks!

Unfortunately the move to php-parser v5 means it now conflicts with phpunit v9 which is the last version to support PHP 7.4 which we're on for at least the next 6 months, so I need to figure out what to do about this.

**Comment by dotgit** [comment]
- `php-parser` v5 works on php 7.4 and up
- it defines itself `phpunit` v9 in composer.json
- `text` works with `phpunit` v9 without an issue

how can they all conflict with each other?

**Comment by cpeel** [comment]
I got this working with the right massaging. [phpunit 9.6.19](https://packagist.org/packages/phpunit/phpunit#9.6.19) (last one to support PHP 7.4) requires [php-code-coverage](https://packagist.org/packages/phpunit/php-code-coverage#9.2.28) which requires nikic/php-parser. The _latest_ php-code-coverage adds support for php-parser v5 but I had to get composer to upgrade things in the right order -- or rather uninstall phpunit and text and re-install them to get php-parser v5.

Good news is that I got installed and the initial look says it's working beautifully. I'll get our eagle-eyed translation guru to test-drive it.

Thank you so much for all of your help! Our French and German users in particular will be very happy to have more of our site correctly translated for them.
